File: src/dstream/hybrid_0_steady_1_tilted_2_algo.zig
const std = @import("std");

const aux = @import("../_auxlib.zig");

const steady_algo = @import("./steady_algo.zig");
const tilted_algo = @import("./tilted_algo.zig");

/// Does this algorithm have the capacity to ingest a data item at logical time
/// T?
///
/// @param u Unsigned integer type for operands and return value.
/// @param S The number of buffer sites available.
/// @param T Queried logical time.
/// @returns Whether there is capacity to ingest at time T.
pub fn has_ingest_capacity(comptime u: type, S: u, T: u) bool {
    aux.assert_unsigned(u);
    const half_S = S >> 1;
    const has_capacity_1st = steady_algo.has_ingest_capacity(
        u,
        half_S,
        T >> 1,
    );
    const has_capacity_2nd = ((T == 0) or tilted_algo.has_ingest_capacity(
        u,
        half_S,
        (T - 1) >> 1,
    ));
    return has_capacity_1st and has_capacity_2nd;
}

/// Site selection for hybrid steady/tilted curation.
///
/// What buffer site should the T'th data item be stored to?
///
/// @param u Unsigned integer type for operands and return value.
/// @param S Buffer size.
///     Must be a power of two, greater than 2.
/// @param T Current logical time.
///    Must be less than 2^(S - 1) - 1.
pub fn assign_storage_site(comptime u: type, S: u, T: u) u {
    aux.assert_unsigned(u);
    std.debug.assert(has_ingest_capacity(u, S, T));

    const half_S = S >> 1;
    const half_T = T >> 1;
    if (T & 1 == 0) {
        const site = steady_algo.assign_storage_site(u, half_S, half_T);
        return if (site == half_S) S else site;
    } else {
        return half_S + tilted_algo.assign_storage_site(u, half_S, half_T);
    }
}

File: src/dstream/hybrid_0_steady_1_tilted_2_algo.rs
use super::{AssignStorageSiteTrait, HasIngestCapacityTrait, SteadyAlgo, TiltedAlgo};
use crate::_auxlib as aux;

/// Does this algorithm have the capacity to ingest a data item at logical time
/// T?
///
/// @template Uint Unsigned integer type for operands.
/// @param S The number of buffer sites available.
/// @param T Queried logical time.
/// @returns Whether there is capacity to ingest at time T.
#[allow(non_snake_case)]
pub fn has_ingest_capacity<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> bool {
    let half_S: Uint = S >> 1;
    let has_capacity_1st: bool = SteadyAlgo::has_ingest_capacity(half_S, T >> 1);

    let has_capacity_2nd =
        T == Uint::zero() || TiltedAlgo::has_ingest_capacity(half_S, (T - Uint::one()) >> 1);

    has_capacity_1st && has_capacity_2nd
}

/// Site selection for hybrid steady/tilted curation.
///
/// What buffer site should the T'th data item be stored to?
///
/// @template Uint Unsigned integer type for operands and return value.
/// @param S Buffer size.
///     Must be a power of two, greater than 2.
/// @param T Current logical time.
///    Must be less than 2^(S - 1) - 1.
/// @returns The selected storage site, if any.
///     Returns S if no site should be selected (i.e., discard).
#[allow(non_snake_case)]
pub fn _assign_storage_site<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> Uint {
    debug_assert!(has_ingest_capacity(S, T));

    let half_S: Uint = S >> 1;
    let half_T: Uint = T >> 1;

    if T & Uint::one() == Uint::zero() {
        let site: Uint = SteadyAlgo::_assign_storage_site(half_S, half_T);
        if site == half_S {
            S
        } else {
            site
        }
    } else {
        half_S + TiltedAlgo::_assign_storage_site(half_S, half_T)
    }
}

/// Site selection for hybrid steady/tilted curation.
///
/// What buffer site should the T'th data item be stored to?
///
/// @template Uint Unsigned integer type for operands and return value.
/// @param S Buffer size.
///     Must be a power of two, greater than 2.
/// @param T Current logical time.
/// @returns The selected storage site, if any.
///     Returns None if no site should be selected (i.e., discard).
#[allow(non_snake_case)]
pub fn assign_storage_site<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> Option<Uint> {
    let k = _assign_storage_site(S, T);
    if k == S {
        None
    } else {
        Some(k)
    }
}

pub struct Algo;

#[allow(non_snake_case)]
impl HasIngestCapacityTrait for Algo {
    fn has_ingest_capacity<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> bool {
        has_ingest_capacity::<Uint>(S, T)
    }
}

#[allow(non_snake_case)]
impl AssignStorageSiteTrait for Algo {
    fn _assign_storage_site<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> Uint {
        _assign_storage_site::<Uint>(S, T)
    }

    fn assign_storage_site<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> Option<Uint> {
        assign_storage_site::<Uint>(S, T)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_smoke_has_ingest_capacity() {
        has_ingest_capacity::<u32>(16, 101);
    }

    #[test]
    fn test_smoke_assign_storage_site() {
        assign_storage_site::<u32>(16, 101);
    }

    #[test]
    fn test_smoke_impl_assign_storage_site() {
        _assign_storage_site::<u32>(16, 101);
    }
}
